### ISSUE TITLE
PyQt5 and glumpy. When closing a window, an error occurs GLError 1281

### ISSUE THREAD

**Issue by Frekby** [glumpy/glumpy#244]
Hello, using the module with pyqt5 and when closing the window there is such an error 
GLError: GLError(
	err = 1281,
	description = b'\xed\xe5\xe4\xee\xef\xf3\xf1\xf2\xe8\xec\xee\xe5 \xe7\xed\xe0\xf7\xe5\xed\xe8\xe5',
	baseOperation = glUseProgram,
	cArguments = (1,)
)
Moreover, when I restart the program, it may not work correctly. And if you first clear the window using window._native_window.close (), then close the pyqt window, then everything is fine. But I can’t do it at the same time to close the qt and glumpy windows

**Comment by rougier** [comment]
If you close the window without telling glumpy (i.e. closing the pyqt window), glumpy may try to draw something and since th GL context has been destroyed with the closing, that might be the reason.

**Comment by Frekby** [comment]
> 
> 
> If you close the window without telling glumpy (i.e. closing the pyqt window), glumpy may try to draw something and since th GL context has been destroyed with the closing, that might be the reason.

Yes I understand this, but I can't close the glumpy and pyqt window at the same time. Although when processing the output event, I first close glumpy and then pyqt. But it doesn't work for some reason

**Comment by melMass** [comment]
I had the same issue, I think it could be added to the backend file directly but this is what I do to solve the issue, you need to close the GL object manually in your Qt close event:

```python
class GlumpyWidget(QMainWindow):
    def __init__(self, parent=None, flags=Qt.WindowFlags()):
        super().__init__(parent=parent, flags=flags)
        self.setWindowTitle(APP_TITLE)
        self.glview = Window(width=700, height=280, color=(0.15, 0.15, 0.15, 1))
        self.setCentralWidget(self.glview)

    def closeEvent(self, ev):
        self.glview.close() # <---- This is needed to avoid the memory crash
        return super().closeEvent(ev)
```

**Comment by Frekby** [comment]
> 
> 
> I had the same issue, I think it could be added to the backend file directly but this is what I do to solve the issue, you need to close the GL object manually in your Qt close event:
> 
> ```python
> class GlumpyWidget(QMainWindow):
>     def __init__(self, parent=None, flags=Qt.WindowFlags()):
>         super().__init__(parent=parent, flags=flags)
>         self.setWindowTitle(APP_TITLE)
>         self.glview = Window(width=700, height=280, color=(0.15, 0.15, 0.15, 1))
>         self.setCentralWidget(self.glview)
> 
>     def closeEvent(self, ev):
>         self.glview.close() # <---- This is needed to avoid the memory crash
>         return super().closeEvent(ev)
> ```

I almost do the same, but return super().closeEvent(ev) without this line. But it doesn't work for me

**Comment by rougier** [comment]
Might be a matter of timing. Can you insert a sleep (a few milliseconds) between `self.glview.close()` and `super().closeEvent(ev)`? If this works, we may need to have a `block` option in the `glview.close` in order to wait for the actual closing.

**Comment by Frekby** [comment]
> 
> 
> Might be a matter of timing. Can you insert a sleep (a few milliseconds) between `self.glview.close()` and `super().closeEvent(ev)`? If this works, we may need to have a `block` option in the `glview.close` in order to wait for the actual closing.

I tried to put sleep to the program in this place, but as I understood Qt modules work streamwise. I have why the first program fell asleep, and then only tried to close the window glumpy

**Comment by rougier** [comment]
You mean it doesn't work ?

**Comment by Frekby** [comment]
> 
> 
> You mean it doesn't work ?

No, I had to make another intermediate window that is called to close the entire program. In other words I first close the grumpy window and open a new window that closes the main one

**Comment by rougier** [comment]
Weird. I'm not familiar enough with Qt to be of any help on this problem. I guess there should be a nice way to close the window but I'm out of idea right now.
